File: Shifting-Letters/SL.rs
pub fn shifting_letters(s: String, shifts: Vec<i32>) -> String {
    let mut sum = 0;
    let mut s_b = s.bytes().rev();
    let mut result: Vec<u8> = vec![];

    for i in shifts.iter().rev() {
        sum += i;
        sum %= 26;
        let mut this_one = s_b.next().unwrap() as i32 + sum;
        if this_one > 122 {
            this_one -= 26;
        }

        result.push(this_one as u8);
    }

    result.reverse();
    String::from_utf8(result).unwrap()
    // match String::from_utf8(result) {
    //     Ok(s) => s,
    //     Err(e) => {
    //         println!("{:?}", e);
    //         String::new()
    //     }
    // }
}

fn main() {
    //println!("{:?}", 97 as u8);
    dbg!(shifting_letters(String::from("abc"), vec![3, 5, 9])); //=> rpl
    dbg!(shifting_letters(String::from("bad"), vec![10, 20, 30])); //=> jyh
    dbg!(shifting_letters(
        String::from("mkgfzkkuxownxvfvxasy"),
        vec![
            505870226, 437526072, 266740649, 224336793, 532917782, 311122363, 567754492, 595798950,
            81520022, 684110326, 137742843, 275267355, 856903962, 148291585, 919054234, 467541837,
            622939912, 116899933, 983296461, 536563513
        ]
    )); //=>
}

File: Shifting-Letters/SL.hs
import           Data.Char

shifting_letters :: [Char] -> [Int] -> [Char]
shifting_letters s l = inner_shift ([], (reverse s)) (0, (reverse l))

inner_shift :: ([Char], [Char]) -> (Int, [Int]) -> [Char]
inner_shift (result, []) (_, _)               = result
inner_shift (result, _) (_, [])               = result
inner_shift (result, (x:xs)) (before, (y:ys)) =
  let (c, e) = shift_char x (before, y)
   in inner_shift ((c : result), xs) (e, ys)

type Before = Int
type ThisInd = Int

shift_char :: Char -> (Before, ThisInd) -> (Char, Before)
shift_char c (ib, i) =
  let new_sum = mod (ib + i) 26
      this_char = ((ord c) + new_sum)
      x = if this_char > 122 then this_char -26 else this_char
   in (chr x, new_sum)


main = do
  print $ shifting_letters "abc" [3, 5, 9]
  print $ shifting_letters "bad" [10, 20, 30]
  print $
    shifting_letters
      "mkgfzkkuxownxvfvxasy"
      [ 505870226
      , 437526072
      , 266740649
      , 224336793
      , 532917782
      , 311122363
      , 567754492
      , 595798950
      , 81520022
      , 684110326
      , 137742843
      , 275267355
      , 856903962
      , 148291585
      , 919054234
      , 467541837
      , 622939912
      , 116899933
      , 983296461
      , 536563513
      ]
